lib/charms/tempo_coordinator_k8s/v0/tempo_api.py

```diff
@@ -0,0 +1,219 @@
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = ["pydantic>=2"]
+
+log = logging.getLogger(__name__)
+
+DEFAULT_RELATION_NAME = "tempo-api"
+
+
+class TempoApiAppData(BaseModel):
+    """Data model for the tempo-api interface."""
+
+    ingress_url: Optional[AnyHttpUrl] = Field(
+        default=None,
+        description="The non-internal URL at which this application can be reached.  Typically, this is an ingress URL.",
+    )
+    direct_url: AnyHttpUrl = Field(
+        description="The cluster-internal URL at which this application can be reached.  Typically, this is a"
+        " Kubernetes FQDN like name.namespace.svc.cluster.local for connecting to the prometheus api"
+        " from inside the cluster, with scheme."
+    )
+
+
+class TempoApiRequirer:
+    """The requirer side of the tempo-api relation."""
+
+    def __init__(
+        self,
+        relations: RelationMapping,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ) -> None:
+        """Initialize the TempoApiRequirer object.
+
+        This object is for accessing data from relations that use the tempo-api interface.  It **does not**
+        autonomously handle the events associated with that relation.  It is up to the charm using this object to
+        observe those events as they see fit.  Typically, that charm should observe this relation's relation-changed
+        event.
+
+        This object is for interacting with a relation that has limit=1 set in charmcraft.yaml.  In particular, the
+        get_data method will raise if more than one related application is available.
+
+        Args:
+            relations: The RelationMapping of a charm (typically `self.model.relations` from within a charm object).
+            relation_name: The name of the relation.
+        """
+        self._charm_relation_mapping = relations
+        self._relation_name = relation_name
+
+    @property
+    def relations(self):
+        """Return the relation instances for applications related to us on the monitored relation."""
+        return self._charm_relation_mapping.get(self._relation_name, ())
+
+    def get_data(self) -> Optional[BaseModel]:
+        """Return data for at most one related application, raising if more than one is available.
+
+        Useful for charms that always expect exactly one related application.  It is recommended that those charms also
+        set limit=1 for that relation in charmcraft.yaml.  Returns None if no data is available (either because no
+        applications are related to us, or because the related application has not sent data).
+        """
+        relations = self.relations
+        if len(relations) == 0:
+            return None
+        if len(relations) > 1:
+            # TODO: Different exception type?
```

Comment:
if we can prevent this altogether by validating in init, that's best IMO

Reply:
I like this idea with one caveat

If someone instantiates this inside their charm init, they need to handle that error otherwise everything in the charm is blocked.  Does that change your mind on what is best?  I'm not sure how I feel about it.

Alternatively, I don't think this should be instantiated in charm.__init__ at all, but that's a separate thread :D

Reply:
Thinking on this more, I still have reservations about this sort of "can fail intermittently at runtime" check in init.  I have a specific example from istio's ingress for why I think its important, but we can discuss offline if interested.

Regardless though, I think this'll get resolved by [this thread](https://github.com/canonical/tempo-coordinator-k8s-operator/pull/119#discussion_r1969323368) if we enforce limit=1

Reply:
I think this is resolved now with the `_validate_relation_metadata` method.  I'll close this if the other thread gets a :+1:

Reply:
agreed, it's not a case of 'failing intermittently at runtime', but a case of 'this will 100% fail all the time unless you do it right, and you should catch this in every unittest in which you are covering this branch'. So the only case in which a charm without limit=1 will make it into live juju is one in which the user isn't writing (good) unittests, so that's on them.

And to be fair, if they don't write unittests, how likely are they to be reading the docs?